### PR TITLE
Reuse expensive merge in theme

### PR DESCRIPTION
**Pull request by nhannah**
This PR build's on @vishalnarkhede previous theme improvement to further reduce the number of merge's / lodash calls made. I introduced a new prop to ThemeProvider that instead of performing a merge uses the Theme directly, for messages and overlay this allows for reuse of the current theme. Essentially all theme merges should now only happen once or when a change occurs instead.

### CODE REVIEW

**Comment by nhannah** [comment]
Note, we don't love objects in dependency arrays but for this case it makes sense as we want to rerun the memoized functions if the pointer changes.

File: src/contexts/themeContext/ThemeContext.tsx [hunk 1]
```diff
@@ -8,26 +8,40 @@ export type DeepPartial<T> = {
...
-    if (style) {
-      merge(finalTheme, style);
+    if (mergedStyle) {
+      return mergedStyle;
     }
 
-    return finalTheme;
-  }, [theme, style]);
+    return mergeThemes({ style, theme });
+  }, [theme, style, mergedStyle]);
```
Comment (virdesai) [inline src/contexts/themeContext/ThemeContext.tsx:44]:
```suggestion
  }, [mergedStyle, style, theme]);
```

File: src/components/MessageOverlay/MessageOverlay.tsx [hunk 2]
```diff
@@ -145,24 +146,40 @@ const MessageOverlayWithContext = <
...
+
+  useEffect(() => {
+    if (myMessageTheme) {
+      setMyMessageThemeString(JSON.stringify(myMessageTheme));
+    }
+  }, [myMessageTheme]);
+
+  const modifiedTheme = useMemo(
+    () => mergeThemes({ style: myMessageTheme, theme }),
+    [theme, myMessageThemeString],
```
Comment (virdesai) [inline src/components/MessageOverlay/MessageOverlay.tsx:181]:
```suggestion
    [myMessageThemeString, theme],
```

File: src/components/MessageList/MessageList.tsx [hunk 3]
```diff
@@ -295,12 +296,17 @@ const MessageListWithContext = <
...
-      messageList: { container, listContainer },
-    },
-  } = useTheme();
+    colors: { accent_blue, white_snow },
+    messageList: { container, listContainer },
+  } = theme;
+
+  const modifiedTheme = useMemo(
+    () => mergeThemes({ style: myMessageTheme, theme }),
+    [theme, myMessageTheme],
```
Comment (virdesai) [inline src/components/MessageList/MessageList.tsx:308]:
```suggestion
    [myMessageTheme, theme],
```

**Review by virdesai — APPROVED** [review]

**Comment by nhannah** [comment]
Haha damn, I literally thought 30min ago, should double back and ABC these arrays then forgot.

**Comment by vishalnarkhede** [comment]
> Haha damn, I literally thought 30min ago, should double back and ABC these arrays then forgot.

line-up-in-alphabetical-order-aron-aron-aaron-qr0S7

**Review by vishalnarkhede — APPROVED** [review]

**Pull request merged**